primitives/utils/src/metrics.rs

```diff
@@ -0,0 +1,35 @@
+
+use lazy_static::lazy_static;
+use prometheus::{
+    Registry, Error as PrometheusError, Opts,
+    core::{
+        GenericGaugeVec, AtomicU64,
+        GenericGauge, GenericCounter
+	}
+};
+
+lazy_static! {
+
+    pub static ref TOKIO_THREADS_TOTAL: GenericCounter<AtomicU64> = GenericCounter::new(
+        "tokio_threads_total", "Total number of threads created"
+    ).expect("Creating of statics doesn't fail. qed");
+
+    pub static ref TOKIO_THREADS_ALIVE: GenericGauge<AtomicU64> = GenericGauge::new(
+        "tokio_threads_alive", "Number of threads alive right now"
+    ).expect("Creating of statics doesn't fail. qed");
+
+	pub static ref UNBOUNDED_CHANNELS_COUNTER : GenericGaugeVec<AtomicU64> = GenericGaugeVec::new(
+        Opts::new("unbounded_channel_len", "Items in each mpsc::unbounded instance"),
```

Comment:
I would suggest having both two counters per channel, one for items send and one for items received. This enables us to see spikes in between scrape intervals. E.g. consider the following time line:

1. Prometheus scrapes the metric endpoint.

2. 1_000_000_000 messages are send into the channel.

3. 1_000_000_000 messages are received from the channel.

4. Prometheus scrapes the metric endpoint.

Reply:
Could just be an `action` label with the possible values of `send` and `received`.

Reply:
This implementation focuses on finding the memory leaks – thus this cares less about the number of messages sent and receive, rather than that there's nothing left unconsumed ... for e.g. because something stalls. Thus this used the combined counter rather than having to combine two graphs to learn about it, but if you tell me that the dashboard then shows the combined currently-alive then I don't care.

I would move the label from "received" to "consumed" though, as we also care about the items leaving memory on-drop, which have never been "received". Or do you think we should have a third counter for that?

Reply:
> Thus this used the combined counter rather than having to combine two graphs to learn about it, but if you tell me that the dashboard then shows the combined currently-alive then I don't care.

If you use labels to differentiate actions they can be displayed easily in a single graph in Prometheus.

> I would move the label from "received" to "consumed" though, as we also care about the items leaving memory on-drop, which have never been "received". Or do you think we should have a third counter for that?

That is a  good point. How about a new label `action` with the values `send` `consumed` (or how about `dropped`?) and `received`?

Reply:
I used `received` and `dropped` already.